File: python/intensify/core/kernels/power_law.py
"""Power-law kernel for Hawkes processes."""

import numpy as np

from .base import Kernel



class PowerLawKernel(Kernel):
    r"""
    Power-law kernel: φ(t) = α (t + c)^{-(1+β)}.

    Parameters
    ----------
    alpha : float
        Amplitude. Must be positive.
    beta : float
        Tail exponent (> 0). Larger beta => heavier tail? Actually larger beta makes decay faster.
        For heavy tail, beta small.
    c : float
        Offset parameter to avoid singularity at t=0. Must be positive (c > 0).

    Notes
    -----
    L1 norm must be computed numerically and may be >= 1 for heavy tails.
    This kernel does NOT have recursive form; uses O(N²) general likelihood.

    Examples
    --------
    >>> kernel = PowerLawKernel(alpha=0.5, beta=0.8, c=0.1)
    """

    def __init__(self, alpha: float, beta: float, c: float = 1.0):
        if alpha <= 0:
            raise ValueError("alpha must be positive")
        if beta <= 0:
            raise ValueError("beta must be positive")
        if c <= 0:
            raise ValueError("c must be positive to avoid singularity at t=0")
        self.alpha = float(alpha)
        self.beta = float(beta)
        self.c = float(c)

    def evaluate(self, t: np.array) -> np.array:
        """
        φ(t) = α (t + c)^{-(1+β)}
        """
        t = np.asarray(t)
        return self.alpha * np.power(t + self.c, -(1.0 + self.beta))

    def integrate(self, t: float) -> float:
        """
        ∫_0^t φ(τ) dτ.

        For power-law, indefinite integral: ∫ (τ+c)^{-(1+β)} dτ = -(1/β) (τ+c)^{-β}.
        So definite integral from 0 to t: (α/β)[ c^{-β} - (t+c)^{-β} ].
        """
        t_val = float(t)
        term0 = self.c ** (-self.beta)
        termt = (t_val + self.c) ** (-self.beta)
        return self.alpha / self.beta * (term0 - termt)

    def integrate_vec(self, t: np.array) -> np.array:
        t = np.asarray(t)
        term0 = self.c ** (-self.beta)
        termt = np.power(t + self.c, -self.beta)
        return (self.alpha / self.beta) * (term0 - termt)

    def l1_norm(self) -> float:
        """
        ∫_0^∞ φ(t) dt = α / β * c^{-β}.
        """
        return self.alpha / self.beta * (self.c ** (-self.beta))

    def has_recursive_form(self) -> bool:
        return False

    def __repr__(self) -> str:
        return f"PowerLawKernel(alpha={self.alpha}, beta={self.beta}, c={self.c})"

File: crates/kernels/src/power_law.rs
//! Power-law Hawkes kernel: φ(t) = α·(t + c)^{-(1+β)} for t > 0.
//!
//! - Indefinite integral: ∫(τ+c)^{-(1+β)} dτ = -(1/β)·(τ+c)^{-β}
//! - Definite integral 0..t: (α/β)·[c^{-β} − (t+c)^{-β}]
//! - L1 norm: (α/β)·c^{-β}  (finite for any β > 0, c > 0; can exceed 1 for heavy tails)
//! - **No recursive form** — likelihood path is O(N²).

use intensify_core::{IntensifyError, Result};

/// Power-law excitation kernel.
///
/// Mirrors `python/intensify/core/kernels/power_law.py`.
#[derive(Debug, Clone, Copy)]
#[cfg_attr(
    feature = "python",
    pyo3::pyclass(module = "intensify._libintensify.kernels", name = "PowerLawKernel")
)]
pub struct PowerLawKernel {
    pub alpha: f64,
    pub beta: f64,
    pub c: f64,
}

impl PowerLawKernel {
    pub fn new(alpha: f64, beta: f64, c: f64) -> Result<Self> {
        if alpha <= 0.0 {
            return Err(IntensifyError::InvalidParam(format!(
                "alpha must be positive; got {alpha}"
            )));
        }
        if beta <= 0.0 {
            return Err(IntensifyError::InvalidParam(format!(
                "beta must be positive; got {beta}"
            )));
        }
        if c <= 0.0 {
            return Err(IntensifyError::InvalidParam(format!(
                "c must be positive (singularity at t=0 otherwise); got {c}"
            )));
        }
        Ok(Self { alpha, beta, c })
    }

    /// φ(t) = α·(t + c)^{-(1+β)}
    #[inline]
    pub fn evaluate(&self, t: f64) -> f64 {
        self.alpha * (t + self.c).powf(-(1.0 + self.beta))
    }

    /// ∫₀^t φ(τ)dτ = (α/β)·[c^{-β} − (t+c)^{-β}]
    #[inline]
    pub fn integrate(&self, t: f64) -> f64 {
        let c_pow = self.c.powf(-self.beta);
        let tc_pow = (t + self.c).powf(-self.beta);
        (self.alpha / self.beta) * (c_pow - tc_pow)
    }

    /// L1 norm = (α/β)·c^{-β}.
    #[inline]
    pub fn l1_norm(&self) -> f64 {
        (self.alpha / self.beta) * self.c.powf(-self.beta)
    }

    /// Multiply α by `factor` (used for stationarity projection).
    #[inline]
    pub fn scale(&mut self, factor: f64) {
        self.alpha *= factor;
    }

    /// PowerLaw has no recursive sufficient statistic.
    #[inline]
    pub fn has_recursive_form(&self) -> bool {
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;

    #[test]
    fn rejects_invalid_params() {
        assert!(PowerLawKernel::new(0.0, 1.0, 0.5).is_err());
        assert!(PowerLawKernel::new(-0.1, 1.0, 0.5).is_err());
        assert!(PowerLawKernel::new(0.5, 0.0, 0.5).is_err());
        assert!(PowerLawKernel::new(0.5, 1.0, 0.0).is_err());
        assert!(PowerLawKernel::new(0.5, 1.0, -0.1).is_err());
    }

    #[test]
    fn evaluate_at_zero() {
        let k = PowerLawKernel::new(0.5, 0.8, 0.5).unwrap();
        // φ(0) = α·c^{-(1+β)}
        assert_relative_eq!(
            k.evaluate(0.0),
            0.5 * 0.5_f64.powf(-1.8),
            max_relative = 1e-15
        );
    }

    #[test]
    fn integral_to_inf_is_l1_norm() {
        let k = PowerLawKernel::new(0.4, 1.2, 0.3).unwrap();
        // For large t, integrate(t) → l1_norm
        assert_relative_eq!(k.integrate(1e9), k.l1_norm(), max_relative = 1e-9);
    }

    #[test]
    fn integral_matches_quadrature() {
        let k = PowerLawKernel::new(0.4, 0.8, 0.5).unwrap();
        let t_max = 5.0;
        let n = 100_000usize;
        let dt = t_max / n as f64;
        let mut quad = 0.0_f64;
        for i in 0..n {
            let a = i as f64 * dt;
            let b = (i + 1) as f64 * dt;
            quad += 0.5 * (k.evaluate(a) + k.evaluate(b)) * dt;
        }
        assert_relative_eq!(k.integrate(t_max), quad, max_relative = 1e-6);
    }
}
